Expose crates_vendor deps to bzlmod

dtolnay opened the pull request:
Without this, I have been unable to update https://github.com/dtolnay/cxx to bzlmod. When I run `bazel run //third-party:vendor` ([this rule](https://github.com/dtolnay/cxx/blob/3c1bcc012f285dc59d39143e658df67659142cd1/third-party/BUILD#L3-L11)) it failed with **"unknown repo 'cargo_bazel.buildifier-linux-amd64' requested from @@rules_rust~0.35.0"**:

```console
ERROR: no such package '@@[unknown repo 'cargo_bazel.buildifier-linux-amd64' requested from @@rules_rust~0.35.0]//file': The repository '@@[unknown repo 'cargo_bazel.buildifier-linux-amd64' requested from @@rules_rust~0.35.0]' could not be resolved: No repository visible as '@cargo_bazel.buildifier-linux-amd64' from repository '@@rules_rust~0.35.0'
ERROR: ~/.cache/bazel/_bazel_david/ebce1d0721fb68dda9c70c0dd1405803/external/rules_rust~0.35.0/crate_universe/private/vendor/BUILD.bazel:3:27: no such package '@@[unknown repo 'cargo_bazel.buildifier-linux-amd64' requested from @@rules_rust~0.35.0]//file': The repository '@@[unknown repo 'cargo_bazel.buildifier-linux-amd64' requested from @@rules_rust~0.35.0]' could not be resolved: No repository visible as '@cargo_bazel.buildifier-linux-amd64' from repository '@@rules_rust~0.35.0' and referenced by '@@rules_rust~0.35.0//crate_universe/private/vendor:buildifier'
ERROR: Analysis of target '//third-party:vendor' failed; build aborted: Analysis failed
```

After this change, it works. My MODULE.bazel can be seen in https://github.com/dtolnay/cxx/pull/1294.

dtolnay commented on crate_universe/private/vendor_utils.bzl:
```diff
@@ -14,10 +14,13 @@ _BUILDIFIER_SHA256S = {
 }
 
 def crates_vendor_deps():
```
CI failure:

```console
##### :bazel: buildifier: found 1 lint issue in your WORKSPACE, BUILD and *.bzl files
crate_universe/private/vendor_utils.bzl:16:1: function-docstring: The function "crates_vendor_deps" has no docstring.
A docstring is a string literal (not a comment) which should be the first statement of a function body (it may follow comment lines).
```

Presumably this is now needed because the function no longer returns void.

illicitonion reviewed (commented):
Thanks for the fix! I think this looks reasonable (though cc @matts1 for a better look) - would it be possible to expand https://github.com/bazelbuild/rules_rust/tree/main/examples/bzlmod/hello_world and https://github.com/bazelbuild/rules_rust/blob/main/.bcr/presubmit.yml and https://github.com/bazelbuild/rules_rust/blob/main/.bazelci/presubmit.yml to exercise the workflow that was broken for you?

illicitonion commented:
https://github.com/bazelbuild/rules_rust/pull/2373 should help with CI

dtolnay commented:
Thanks!

I've added the `crates_vendor` usage in the **first** commit and implemented the fix in the **second** commit, so you can run `bazel run //third-party:vendor` in the first commit to see that it fails.

```console
ERROR: no such package '@@[unknown repo 'cargo_bazel.buildifier-linux-amd64' requested from @@rules_rust~override]//file': The repository '@@[unknown repo 'cargo_bazel.buildifier-linux-amd64' requested from @@rules_rust~override]' could not be resolved: No repository visible as '@cargo_bazel.buildifier-linux-amd64' from repository '@@rules_rust~override'
ERROR: f2e42d6afe2730a9d87691f9ba152a3d/external/rules_rust~override/crate_universe/private/vendor/BUILD.bazel:3:27: no such package '@@[unknown repo 'cargo_bazel.buildifier-linux-amd64' requested from @@rules_rust~override]//file': The repository '@@[unknown repo 'cargo_bazel.buildifier-linux-amd64' requested from @@rules_rust~override]' could not be resolved: No repository visible as '@cargo_bazel.buildifier-linux-amd64' from repository '@@rules_rust~override' and referenced by '@@rules_rust~override//crate_universe/private/vendor:buildifier'
ERROR: Analysis of target '//third-party:vendor' failed; build aborted: Analysis failed
INFO: Elapsed time: 0.350s, Critical Path: 0.00s
INFO: 1 process: 1 internal.
ERROR: Build did NOT complete successfully
ERROR: Build failed. Not running target
```

dtolnay commented on examples/bzlmod/hello_world/third-party/extension.bzl:
```diff
@@ -0,0 +1,14 @@
+"""Module extensions for using vendored crates with bzlmod"""
+
+load("//third-party/crates:defs.bzl", _crate_repositories = "crate_repositories")
+
+def _crate_repositories_impl(module_ctx):
+    direct_deps = _crate_repositories()
+    return module_ctx.extension_metadata(
+        root_module_direct_deps = [repo.repo for repo in direct_deps],
+        root_module_direct_dev_deps = [],
+    )
+
+crate_repositories = module_extension(
+    implementation = _crate_repositories_impl,
+)
```
Separately, we should make crate_universe automatically generate a suitable module extension. Nothing in this file is project-specific.

illicitonion replied:
I think https://github.com/bazelbuild/rules_rust/issues/2181 is probably the umbrella issue for this... cc @matts1

dtolnay commented on .bcr/presubmit.yml:
```diff
@@ -7,5 +7,7 @@ bcr_test_module:
       working_directory: examples/bzlmod/hello_world
       name: "Run test module"
       platform: ${{ platform }}
+      run_targets:
+        - "//third-party:vendor"
```
In CI this is succeeding on Linux (_"bzlmod BCR presubmit on :ubuntu: 20.04 LTS (OpenJDK 11, gcc 9.4.0)"_) but failing on macOS (_"bzlmod BCR presubmit on :darwin: (OpenJDK 11, Xcode)"_).

```console
/private/var/tmp/_bazel_buildkite/4894c9dc50e84673867cbd80d0914a01/external/rules_rust~override~rust~rust_host_tools/bin/cargo: /private/var/tmp/_bazel_buildkite/4894c9dc50e84673867cbd80d0914a01/external/rules_rust~override~rust~rust_host_tools/bin/cargo: cannot execute binary file

ERROR: /Users/buildkite/builds/bk-imacpro-7/bazel/rules-rust-rustlang/examples/bzlmod/hello_world/third-party/BUILD.bazel:3:14: //third-party:vendor depends on @@rules_rust~override~cargo_bazel_bootstrap~cargo_bazel_bootstrap//:binary in repository @@rules_rust~override~cargo_bazel_bootstrap~cargo_bazel_bootstrap which failed to fetch. no such package '@@rules_rust~override~cargo_bazel_bootstrap~cargo_bazel_bootstrap//': Process exited with code '126'
```

illicitonion replied:
It looks like it's trying to use a Linux `cargo` binary to build `cargo-bazel` - I repro locally... See https://github.com/bazelbuild/rules_rust/pull/2021#issuecomment-1871381693

cc @matts1 @scentini

illicitonion replied:
It looks like the `MODULE.bazel.lock` contains platform-specific information, which gets re-used if nothing else has changed in the repo:

```
          "rust_host_tools": {
            "bzlFile": "@@rules_rust~override//rust:repositories.bzl",
            "ruleClassName": "rust_toolchain_tools_repository",
            "attributes": {
              "name": "rules_rust~override~rust~rust_host_tools",
              "exec_triple": "x86_64-unknown-linux-gnu",
              "target_triple": "x86_64-unknown-linux-gnu",
              "allocator_library": "@rules_rust//ffi/cc/allocator_library",
              "dev_components": false,
              "edition": "2021",
              "rustfmt_version": "nightly/2023-12-07",
              "sha256s": {},
              "urls": [
                "https://static.rust-lang.org/dist/{}.tar.gz"
              ],    
              "version": "1.74.1"
            }     
          },
```

illicitonion replied:
Looks like we need to do something like this: https://github.com/bazelbuild/rules_python/pull/1433/files

dtolnay replied:
Good find. See also https://github.com/bazelbuild/rules_python/issues/1643. Otherwise, according to https://github.com/bazelbuild/rules_python/pull/1433#issuecomment-1772352358, you and all downstream users need to regenerate their lockfiles separately from each platform they support. For example I, on Linux, would not be able to generate a lockfile that passes checks on macOS.

illicitonion replied:
Started a thread on Slack: https://bazelbuild.slack.com/archives/C014RARENH0/p1703787329091939 - hopefully there's a recommendation we can follow :)

illicitonion replied:
I tested this in https://github.com/bazelbuild/rules_rust/pull/2374 - it works, but looks like it's going to churn the lockfile every time someone switches platform. Hopefully we can get some useful pointers soon :)

illicitonion replied:
Yeah, for now let's just delete + .gitignore the `MODULE.bazel.lock` file in examples/bzlmod/hello_world and be happy

matts1 replied:
I believe there's a way  to ensure that you don't use the lockfile for specific module extensions / repository rules. Don't know how though. @Wyverald should know, he implemented the lockfile.

dtolnay commented:
Rebased on #2373 + cherry-picked the lockfile deletion from #2374.

illicitonion reviewed (approved):
Thanks!

@matts1 Can you take a look when convenient and follow up with anything you'd like added/changed? :)

illicitonion commented:
Aah sorry we accidentally merged something else while CI was churning - if you can do one more rebase we can get this merged!
